### ISSUE TITLE
get_class() === ::class

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#228]
The [`get_class()`](http://php.net/manual/en/function.get-class.php) returns current class name from argumented object (else it throw an error). Compare it with the `::class` is lower performance than use `instance_of`.

```php
// Warning: replace it with $object instanceof Object.
get_class($object) === Object::class;
```

**Comment by kalessil** [comment]
For future people: only for interfaces, abstract and final classes.
Regular classes and traits comparison must not be reported (traits are not applicable here, but nevertheless).

**Comment by funivan** [comment]
Interesting thing with `get_class`. May help someone in future =) 
```php
class A {
    public function test(\stdClass $object = null){
      var_dump(get_class($object));
    }
  }

 (new A())->test(); # A
 (new A())->test(new stdClass()); # stdClass
```

**Comment by kalessil** [comment]
WTF, oh my

**Comment by bazzik** [comment]
Weird (stupid if you like) but documented behaviour with `null` as parameter. [Example 2](http://uk3.php.net/manual/en/function.get-class.php#function.usort.examples.multi) from `get_class()` documentation raises more "???" questions as for me...

**Comment by rentalhost** [comment]
Exactly! 

Even, I think that you can implements an inspection to `get_class()` without parameter to PHP < 7.2, because on [this RFC](https://wiki.php.net/rfc/get_class_disallow_null_parameter) it was definetively disallowed. To get current class, you have two options: `get_class($this)` or statically `static::class`. Maybe should I open a new ticket to this?

**Comment by funivan** [comment]
@rentalhost it is not so hard to find `get_class` without parameters.  It is really hard to find null parameter 
```php
$a = User::getData();
if(someMethod()){
  $a = new \stdClass();
}
get_class($a);
```
You need to check all stack and find if `$a` can be null

**Comment by kalessil** [comment]
if I get this correctly, `get_class()` behaves as `__CLASS__`: https://3v4l.org/B56Tc

**Comment by mnpenner** [comment]
I think so. And `get_class($this)` appears to be like `get_called_class()`: https://3v4l.org/Of3GP

**Comment by iltar** [comment]
`static::class` is also an option if you want to get the "current" class afaik.

**Comment by kalessil** [comment]
That's great addition, thank you @mnpenner 

What if we extend `::class constant can be used` inspection then?
- `get_class($this)` -> `static::class`
- `get_called_class()` -> `static::class`

**Comment by mnpenner** [comment]
@kalessil That sounds good to me, but we have to watch out for weirdness. HHVM and PHP behave differently [here](https://3v4l.org/gom2N) -- `get_class($this)` does not behave the same as `static::class` but `get_called_class()` does.

```php
<?php

class A {
    public function test(\stdClass $object = null){
      var_dump(get_class($object));
      var_dump(get_class());
      var_dump(__CLASS__);
      var_dump(get_class($this));
      var_dump(get_called_class());
      var_dump(static::class);
    }
}

class B extends A {
  
}

$rm = new ReflectionMethod(A::class, 'test');
$b = new B;
$rm->invoke($b);
```

**Comment by kalessil** [comment]
@mnpenner : let's stick to `get_called_class()` -> `static::class` then.

**Comment by rentalhost** [comment]
@kalessil is EA compatible with/supports HHVM? Because if not, then you could consider only the PHP result for thinkgs like that.

**Comment by kalessil** [comment]
@rentalhost : depends. 

In this case, we had valuable feedback from @mnpenner, who also mentioned HHVM. We need to take care about HHVM in this case.

If HHVM would not be mentioned, we'd go with the both new patterns for the ::class inspection

**Comment by mnpenner** [comment]
I don't personally use HHVM, but I wouldn't recommend any code mods that make it incompatible. Who knows, maybe the PHP team will employ a similar optimisation that subtly breaks your code because of this.
